### PR TITLE
Deferred Rendering

### PR DESCRIPTION
**Pull request by ttanatb**
## Feature
Graphics pipeline now does deferred rendering

## Implementation/Solution
 - Refactored commandlist construction to fit our data 
 - Reconfigured dsv/rtv/pso's for deferred rendering
 - Died a little inside

## Screenshots/GIF (if applicable)
giphy

## Tests
 - [x] Have you reviewed your own code?
 - [ ] Does it pass unit tests?

### CODE REVIEW

File: Light Vox Engine/Shaders/ps_lighting.hlsl [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+Texture2D gAlbedoTexture    : register(t0);
+Texture2D gNormalTexture    : register(t1);
+Texture2D gPositionTexture  : register(t2);
+
+//Sampler
+SamplerState gSampler       : register(s0);
+
+struct DirectionalLight
+{
+    float4 diffuseColor;
```
Comment (BenjaFriend) [inline Light Vox Engine/Shaders/ps_lighting.hlsl]: Why does the color need to be a `float4`? Isn't it just RGB?

File: Light Vox Engine/ObjLoader.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	DirectX::XMFLOAT3 normal;
+};
+
+struct MeshData
+{
+    std::vector<Vertex> vertices;
+    std::vector<UINT> indices;
+};
+/*
+struct ObjectData
```
Comment (BenjaFriend) [inline Light Vox Engine/ObjLoader.h]: Can this comment be removed?

File: Light Vox Engine/Graphics/GraphicsBase.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    fenceEvent = CreateEvent( nullptr, FALSE, FALSE, nullptr );
+    if ( fenceEvent == nullptr )
+        ThrowIfFailed( HRESULT_FROM_WIN32( GetLastError() ) );
+
+    return S_OK;
+}
+
+HRESULT GraphicsCore::InitFrameResources()
+{
+    for ( int i = 0; i < LV_FRAME_COUNT; i++ )
```
Comment (BenjaFriend) [inline Light Vox Engine/Graphics/GraphicsBase.cpp]: should be `++i`

File: Light Vox Engine/Graphics/GraphicsBase.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+HRESULT GraphicsCore::InitSynchronizationObjects()
+{
+    ThrowIfFailed( device->CreateFence(
+        fenceValue,
+        D3D12_FENCE_FLAG_NONE,
+        IID_PPV_ARGS( &fence )
+    ) );
+    NAME_D3D12_OBJECT( fence );
+    fenceValue++;
```
Comment (BenjaFriend) [inline Light Vox Engine/Graphics/GraphicsBase.cpp]: `++` before, not after

File: Light Vox Engine/Graphics/FrameResource.h [hunk 5]
```diff
@@ -76,34 +87,55 @@ class FrameResource
...
-    Microsoft::WRL::ComPtr<ID3D12PipelineState> pso;
+    Microsoft::WRL::ComPtr<ID3D12PipelineState> geometryBufferPso;
+    Microsoft::WRL::ComPtr<ID3D12PipelineState> deferredPso;
 
     //the gpu side of the scene constant buffer
     Microsoft::WRL::ComPtr<ID3D12Resource> sceneConstantBuffer;
-
-    //instance upload buffer
     Microsoft::WRL::ComPtr<ID3D12Resource> instanceUploadBuffer;
+    //Microsoft::WRL::ComPtr<ID3D12Resource> lightConstantBuffer;
```
Comment (BenjaFriend) [inline Light Vox Engine/Graphics/FrameResource.h]: remove this comment if we don't need it

**Review by BenjaFriend — CHANGES_REQUESTED** [review]

File: Light Vox Engine/Graphics/FrameResource.h (hunk 5 above)
Comment (ttanatb) [reply Light Vox Engine/Graphics/FrameResource.h]: We'll need lights eventually so I'll keep it there for now

File: Light Vox Engine/Graphics/FrameResource.cpp [hunk 6]
```diff
@@ -3,57 +3,254 @@
...
+        //Texture Description for G Buffer
+        D3D12_RESOURCE_DESC textureDesc = { };
+        textureDesc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+        textureDesc.Alignment = 0;
+        textureDesc.SampleDesc.Count = 1;
+        textureDesc.SampleDesc.Quality = 0;
+        textureDesc.MipLevels = textureMipLevels;
+        textureDesc.DepthOrArraySize = 1;
+        textureDesc.Width = viewport->Width;
+        textureDesc.Height = viewport->Height;
```
Comment (ArturoKuang) [inline Light Vox Engine/Graphics/FrameResource.cpp]: implicit conversion float to uint

**Review by ArturoKuang — DISMISSED** [review]

**Review by BenjaFriend — APPROVED** [review]

**Pull request merged**